src/lib/uri/uri.c

```diff
@@ -7,11 +7,172 @@
 #include "parser.h"
 #include "trivia/util.h"
 
-#define XSTRNDUP(s, n) (s != NULL ? xstrndup(s, n) : NULL)
+#define XSTRNDUP(s, n)  (s != NULL ? xstrndup(s, n) : NULL)
+#define CONTAINER_INCREASE_SIZE(container, name) \
+({ \
+        __typeof__(container->name) *array = &container->name; \
+        int *count = &container->name##_##count; \
+        unsigned long elem_size = sizeof(__typeof__((*array)[0])); \
+        *array = xrealloc(*array, (*count + 1) * elem_size); \
+        __typeof__ ((*array)[0]) *elem = &((*array)[(*count)++]); \
+        memset(elem, 0, elem_size); \
+        elem; \
+})
+
+/**
+ * Found URI option @a uri_options array by it's @a name.
+ */
+static struct uri_option *
+uri_option_by_name(const struct uri_options *uri_options, const char *name)
+{
+        for (int i = 0; i < uri_options->options_count; i++)
+                if (strcmp(uri_options->options[i].name, name) == 0)
+                        return &uri_options->options[i];
+        return NULL;
+}
+
+/**
+ * Found URI option in @a uri_options array, by string, which contains
+ * option name and option values separated by '='.
+ */
+static struct uri_option *
+uri_option_by_name_with_separator(struct uri_options *uri_options, char *option)
+{
+        char *opt_delim = strchr(option, '=');
+        if (opt_delim != NULL)
+                *opt_delim = '\0';
+        struct uri_option *uri_option = uri_option_by_name(uri_options, option);
+        if (opt_delim)
+                *opt_delim = '=';
+        return uri_option;
+}
+
+/**
+ * Destroy @a option_value and all associated resources.
+ */
+static void
+uri_option_value_destroy(struct uri_option_value *option_value)
+{
+        free(option_value->value);
+}
+
+/**
+ * Create @a option_value form @a value string
+ */
+static void
+uri_option_value_create(struct uri_option_value *option_value,
+                        const char *value)
+{
+        option_value->value = xstrdup(value);
+}
+
+/**
+ * Destroy @a values and all associated resources.
+ */
+static void
+uri_option_values_destroy(struct uri_option_values *values)
+{
+        for (int i = 0; i < values->values_count; i++)
+                uri_option_value_destroy(&values->values[i]);
+        free(values->values);
+        for (int i = 0; i < values->copies_count; i++)
+                free(values->copies[i]);
+        free(values->copies);
+        memset(values, 0, sizeof(struct uri_option_values));
+}
+
+/**
+ * Creates @a option_values form @a values string. Expected @a values
+ * format is a string which contains option values separated by ';'.
+ * For example: "10;20;30".
```

Comment:
`;` is typically used instead of `&` to separate query parameters. Even if we aren't going to support it, it's bad to reuse it for separating array members IMO. A comma (`,`) looks like a more standard way, but I'm not sure we can use it, because we're planning to use it as a separator for URIs... What about `a[]=x;a[]=y;a[]=z`? Please google other ways typically used for passing arrays in query parameters.